cirq-core/cirq/qis/clifford_tableau.py

```diff
@@ -176,6 +179,67 @@ def _str_full_(self) -> str:
 
         return string
 
+    def then(self, second: 'CliffordTableau') -> 'CliffordTableau':
+        """Returns a composed CliffordTableau of this tableau and the second tableau.
+
+        Then composed tableau is equal to (up to global phase) the composed
+        unitary operation of the two tableaux, i.e. equivalent to applying the unitary
+        operation of this CliffordTableau then applying the second one.
+
+        Args:
+            second: The second CliffordTableau to compose with.
+
+        Returns:
+            The composed CliffordTableau.
+
+        Raises:
+            TypeError: If the type of second is not CliffordTableau.
+            ValueError: If the number of qubits in the second tableau mismatch with
+                this tableau.
+        """
+        if not isinstance(second, CliffordTableau):
+            raise TypeError("The type for second tableau must be the CliffordTableau type")
+        if self.n != second.n:
+            raise ValueError(
+                f"Mismatched number of qubits of two tableaux: {self.n} vs {second.n}."
+            )
+
+        # Convert the underlying data type from bool to int for easier numerical computation.
+        m1 = self.matrix().astype(int)
+        m2 = second.matrix().astype(int)
+
+        p1 = self.rs.astype(int)
+        p2 = second.rs.astype(int)
+
+        merged_m = np.mod(m1.dot(m2), 2)
+        phase = np.mod(p1 + m1.dot(p2), 2)
+
+        # we need more phase correction for expanding Y to XZ and swapping Z_iX_i order.
+        for k in range(2 * self.n):
+            swap_phase = 0  # value betwen 0 and 3 representing [1, i, -1, -i] respectively.
+            prev_row_sum = np.zeros([2 * self.n])
+            swap_phase += np.sum(m1[k, : self.n] * m1[k, self.n :])  # Y gate => iXZ
+            for i, v in enumerate(m1[k]):
+                if v == 0:
+                    continue
+                swap_phase += np.sum(m2[i, : self.n] * m2[i, self.n :])  # Y gate => iXZ
+                swap_phase += 2 * np.sum(m2[i, : self.n] * prev_row_sum[self.n :])  # swap Z_i X_i
+                prev_row_sum += m2[i]
+            prev_row_sum = np.mod(prev_row_sum, 2)
+            swap_phase -= np.sum(prev_row_sum[: self.n] * prev_row_sum[self.n :])  # XZ => -iY
+            phase[k] = (phase[k] + (swap_phase % 4) / 2) % 2
```

Comment:
Can you add more explanatory comments to these lines? A reference would be great too. It is really not obvious to me that why we are adding observables (rows) together for example to count the Ys..

Reply:
I get the %4, counting 1j phases for each ZX and the 1j * 1j for -1 for XZ. But the way we combine the second tableaux is just hard to parse from the code itself, which I think should be a criteria for readable code, or at least we should have a ref, so that someone can understand it from there. The original Aaronson and Gottesman paper did not talk about combining the tableaux.

Reply:
Sure, add more sentences. But it is really hard for me to describe it precisely. The key idea is global phase is the only scalar so that we can always move that to the beginning. So we can safely compute the phase of each component whenever we like then add them at last. We have lots of components to generate phases `(phase[k] + (swap_phase % 4) / 2) % 2` the first `phase[k]` is the phase we generate by composing the two tableaux without adjusting the order etc(note like origin table may have +/- phase for each stabilizer). 

I don't know a good reference for it. @Strilanc Do you know any good references for it?

Reply:
So, I found this, where Theorem 36 describes a vectorized solution: https://arxiv.org/pdf/2009.03218.pdf, it was so small, I implemented it, but the code is not working yet, it fails on some of the random circuit tests, I probably made a mistake somewhere, at least I thoroughly reviewed your tests, and they do seem to be sound :)  

```
       m1 = self.matrix().astype(int)
        m2 = second.matrix().astype(int)

        lmbda = np.zeros((2 * self.n, 2 * self.n))
        lmbda[:self.n, self.n:] = np.eye(self.n)
        p1 = np.diag(m1 @ lmbda @ m1.T)
        m2Lm2T = m2 @ lmbda @ m2.T
        p2 = np.diag(m2Lm2T)

        s1 = self.rs.astype(int)
        s2 = second.rs.astype(int)

        merged_m = np.mod(m1 @ m2, 2)
        p12 = np.mod(np.diag(merged_m @ lmbda @ merged_m.T), 2)
        assert np.allclose(p12, np.mod(p1 + m1 @ p2, 2)), f"expected: {p12}, got: {p1 + m1 @ p2}"

        signs = np.mod(s1 + p1 * (m1 @ p2) + m1 @ s2 + np.diag(
            m1 @ np.tril(np.outer(p2, p2.T) + m2Lm2T, -1) @ m1.T), 2)

        merged_tableau = CliffordTableau(num_qubits=self.n)
        merged_tableau.xs = merged_m[:, :self.n]
        merged_tableau.zs = merged_m[:, self.n:]
        merged_tableau.rs = signs
```

Reply:
Thanks for finding this reference! I spent some time figuring out how to implement it correctly. In coding, it is not as clean as the equations. The ugly part is they use `i^a(-1)^bX^cZ^d` format, where a,b,c,d is a boolean type. We need to convert to this format first, do the computation in that domain, then convert it back. In both conversions, counting the number of y gates is inevitable. Check this modification based on your code:
```
        m1 = self.matrix().astype(int)
        m2 = second.matrix().astype(int)

        num_ys1 = np.sum(m1[:, : self.n] * m1[:, self.n :], axis=1)
        num_ys2 = np.sum(m2[:, : self.n] * m2[:, self.n :], axis=1)

        p1 = np.mod(num_ys1, 2)
        p2 = np.mod(num_ys2, 2)

        s1 = self.rs.astype(int) + np.mod(num_ys1, 4) // 2
        s2 = second.rs.astype(int) + np.mod(num_ys2, 4) // 2

        lmbda = np.zeros((2 * self.n, 2 * self.n))
        lmbda[: self.n, self.n :] = np.eye(self.n)
        m2Lm2T = m2 @ lmbda @ m2.T

        m_12 = np.mod(m1.dot(m2), 2)
        p_12 = np.mod(p1 + m1.dot(p2), 2)
        s_12 = (
            s1
            + m1.dot(s2)
            + p1 * m1.dot(p2)
            + np.diag(m1 @ np.tril(np.outer(p2, p2.T) + m2Lm2T, -1) @ m1.T)
        )
        num_ys12 = np.sum(m_12[:, : self.n] * m_12[:, self.n :], axis=1)
        merged_phase = np.mod(p_12 + 2*s_12 - num_ys12, 4)
        merged_m = m_12

        merged_tableau = CliffordTableau(num_qubits=self.n)
        merged_tableau.xs = merged_m[:, : self.n]
        merged_tableau.zs = merged_m[:, self.n :]
        merged_tableau.rs = merged_phase
```

Reply:
Very nice! I'm impressed how you deeply understood all of this :) 
You are right, the Aaronson & Gottesman representation's `r` is not the same as the `s` vector in the Gosset paper's representation. In order to calculate with the Gosset method, we need to "remove" the "signs" that come from the PauliStrings that have even number of Ys in them. 

I'd be curious to see which implementation is faster. I started doing a bit of benchmarking by modifying the end of your `then` test:

```
    n_qubits = 400
    t_final, t_op, expected_t = _three_identical_table(n_qubits)
    seq_op = random_circuit(num_ops=1000, num_qubits=n_qubits)
    for i, (op, args) in enumerate(seq_op):
            t_op = cirq.CliffordTableau(n_qubits)
            op(t_op, *args)
            t_final = t_final.then(t_op)
            op(expected_t, *args)

    assert expected_t == t_final
```

And started to measure based on n_qubits. 
On my machine 

qubits | new | old
-- | -- | --
100 | 6.56 | 30.8
200 | 61.8 | 105.74
300 | 204.04 | 276.61
400 | 511.63 | 670

Which seems like the first version is slower at first blink, but it seems to be growing slower (with an exponent of ~2 instead of ~3.2)...but we need more datapoints. Both should be O(n^3), so these measurements are probably heavily influenced by vectorization and constant factors. 

So - I like the Gosset version a bit more because we have a nice paper + proof describing it and it looks vectorized, so it should be faster - but we need to confirm it. I think speed does matter here, if anybody will use this for QEC, they will use large tableaux.

Reply:
Just to close this off, the new method is slightly faster but not significantly - my previous data was messed up with manual datataking - this one was automated:

qubits | new | old
-- | -- | --
100 | 24.93 | 30.93
200 | 163.97 | 147.83
300 | 350.11 | 378.99
400 | 807.15 | 928.34
500 | 1679.39 | 1776.07
1000 | 18750.73 | 19171.23